### PR TITLE
feat(edge/stacks): check for stacks status [EE-5554]

### PR DESCRIPTION
**Pull request by chiptus**
related to [EE-5554]

[EE-5554]: https://portainer.atlassian.net/browse/EE-5554?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: edge/stack/stack.go [hunk 1]
```diff
@@ -521,6 +610,10 @@ func buildEnvVarsForDeployer(envVars []portainer.Pair) []string {
 }
 
 func (manager *StackManager) deleteStack(ctx context.Context, stack *edgeStack, stackName, stackFileLocation string) {
+	manager.mu.Lock()
```
Comment (andres-portainer) [inline edge/stack/stack.go:616]: Why is this lock needed here?

File: exec/kubernetes_deploy_status.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	"context"
+
+	libstack "github.com/portainer/portainer/pkg/libstack"
+)
+
+func (service *KubernetesDeployer) WaitForStatus(ctx context.Context, name string, status libstack.Status) <-chan string {
+	resultCh := make(chan string)
+
+	go func() {
+		resultCh <- ""
```
Comment (andres-portainer) [inline exec/kubernetes_deploy_status.go]: There is no need to write to the channel here, the `close()` will achieve the same effect.

File: nomad/nomad_deployer.go [hunk 3]
```diff
@@ -255,3 +256,14 @@ func isUpdateJob(job *nomadapi.Job) bool {
 		len(job.TaskGroups[0].Tasks) > 0 &&
 		job.TaskGroups[0].Tasks[0].Name == targetJobName
 }
+
+func (service *Deployer) WaitForStatus(ctx context.Context, name string, status libstack.Status) <-chan string {
+	resultCh := make(chan string)
+
+	go func() {
+		resultCh <- ""
```
Comment (andres-portainer) [inline nomad/nomad_deployer.go]: Same as https://github.com/portainer/agent/pull/478/files#r1254914543.

File: exec/docker_swarm_stack_status.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/portainer/agent/docker"
+	libstack "github.com/portainer/portainer/pkg/libstack"
+	"github.com/rs/zerolog/log"
+)
+
+func (service *DockerSwarmStackService) WaitForStatus(ctx context.Context, name string, status libstack.Status) <-chan string {
+	errorMessageCh := make(chan string)
+
+	go func() {
+		timeout := time.After(5 * time.Minute)
```
Comment (andres-portainer) [inline exec/docker_swarm_stack_status.go]: Why do we need the timeout if we (can) have it in the context?

File: exec/docker_swarm_stack_status_test.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+	statusCh := deployer.WaitForStatus(ctx, stackName, requiredStatus)
+	select {
+	case result := <-statusCh:
+		if result == "" {
+			return requiredStatus, "", nil
+		}
+
+		return libstack.StatusError, result, nil
+	case <-ctx.Done():
```
Comment (andres-portainer) [inline exec/docker_swarm_stack_status_test.go]: This channel read contends with the the same one in `deployer.WaitForStatus()`, only one can succeed, I suggest removing it from here.

File: exec/docker_swarm_stack_status.go [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+		for {
+			select {
+			case <-timeout:
+				return
+			case <-ctx.Done():
+				return
```
Comment (andres-portainer) [inline exec/docker_swarm_stack_status.go]: Shouldn't we write to `resultCh` before returning?

**Review by andres-portainer — APPROVED** [review]

File: edge/stack/stack.go (hunk 1 above)
Comment (chiptus) [reply edge/stack/stack.go:616]: because we're changing the stack status. I thought it's needed

File: exec/docker_swarm_stack_status.go (hunk 4 above)
Comment (chiptus) [reply exec/docker_swarm_stack_status.go]: right, it's left from before I added timeout to context

File: exec/docker_swarm_stack_status_test.go (hunk 5 above)
Comment (chiptus) [reply exec/docker_swarm_stack_status_test.go]: :+1:

File: edge/stack/stack.go (hunk 1 above)
Comment (andres-portainer) [reply edge/stack/stack.go:616]: It was needed for this line that is not there anymore
```go
delete(manager.stacks, edgeStackID(stack.ID))
```

I'm not seeing any mutation to the `manager` state here, but I could be missing it.

Comment (chiptus) [reply edge/stack/stack.go:616]: `stack *edgeStack` is an object from the manager state, so changing it means changing the manager state. isn't it?

**Pull request merged**